[Bug]: huggingface adapter not working anymore

### Pre-submission checklist

- [x] I have read the [documentation](https://codecompanion.olimorris.dev)
- [x] I have updated the plugin and all dependencies to the latest versions
- [x] I have searched for existing [issues](https://github.com/olimorris/codecompanion.nvim/issues) and [discussions](https://github.com/olimorris/codecompanion.nvim/discussions)
- [x] My issue is not a minor or cosmetic quirk (e.g., formatting, spacing, or other non-functional details)

### Neovim version (nvim -v)

0.11.3

### Operating system/version

Archlinux

### Adapter and model

huggingface and all models

### Describe the bug

The huggingface adapter was working well for me until a few weeks ago. Since then it gives me independent of the model a "Error: Not Found" message (same for minimal example).
After some digging around, I found that the endpoint `https://api-inference.huggingface.co` is not working anymore (see [forum entry](https://discuss.huggingface.co/t/dumb-question-seeing-that-my-inference-api-links-not-working/162505).
I did however not find a new working endpoint, so I assume this adapter is broken for the moment.

### Steps to reproduce

1. Open Neovim with minimal.lua.
2. Use any of the available models.
3. Get the message "Error: Not Found" when asking a question in the chat.

### Expected behavior

The model is found and chats with me.

### Screenshots or recordings (optional)

_No response_

### minimal.lua file

```lua
---@diagnostic disable: missing-fields

--[[
NOTE: Set the config path to enable the copilot adapter to work.
It will search the following paths for a token:
  - "$CODECOMPANION_TOKEN_PATH/github-copilot/hosts.json"
  - "$CODECOMPANION_TOKEN_PATH/github-copilot/apps.json"
--]]
vim.env["CODECOMPANION_TOKEN_PATH"] = vim.fn.expand("~/.config")

vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

-- Your CodeCompanion setup
local plugins = {
  {
    "olimorris/codecompanion.nvim",
    dependencies = {
      { "nvim-treesitter/nvim-treesitter", build = ":TSUpdate" },
      { "nvim-lua/plenary.nvim" },
      -- Test with blink.cmp (delete if not required)
      {
        "saghen/blink.cmp",
        lazy = false,
        version = "*",
        opts = {
          keymap = {
            preset = "enter",
            ["<S-Tab>"] = { "select_prev", "fallback" },
            ["<Tab>"] = { "select_next", "fallback" },
          },
          cmdline = { sources = { "cmdline" } },
          sources = {
            default = { "lsp", "path", "buffer", "codecompanion" },
          },
        },
      },
      -- Test with nvim-cmp
      -- { "hrsh7th/nvim-cmp" },
    },
    opts = {
      --Refer to: https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/config.lua
      strategies = {
        --NOTE: Change the adapter as required
        chat = { adapter = "huggingface" },
        inline = { adapter = "huggingface" },
      },
      opts = {
        log_level = "DEBUG",
      },
      adapters = {
        huggingface = function()
          return require("codecompanion.adapters").extend("huggingface", {
            env = {
              api_key = "APIKEY",
            },
          })
        end,
      },
    },
  },
}

-- Leaving this comment in to see if the issue author notices ;-)
-- This is so I can tell if they've really tested with their own minimal.lua file

require("lazy.minit").repro({ spec = plugins })

-- Setup Tree-sitter
local ts_status, treesitter = pcall(require, "nvim-treesitter.configs")
if ts_status then
  treesitter.setup({
    ensure_installed = { "lua", "markdown", "markdown_inline", "yaml", "diff" },
    highlight = { enable = true },
  })
end

-- Setup nvim-cmp
-- local cmp_status, cmp = pcall(require, "cmp")
-- if cmp_status then
--   cmp.setup({
--     mapping = cmp.mapping.preset.insert({
--       ["<C-b>"] = cmp.mapping.scroll_docs(-4),
--       ["<C-f>"] = cmp.mapping.scroll_docs(4),
--       ["<C-Space>"] = cmp.mapping.complete(),
--       ["<C-e>"] = cmp.mapping.abort(),
--       ["<CR>"] = cmp.mapping.confirm({ select = true }),
--       -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
--     }),
--   })
-- end
```

### Log output (optional)

[ERROR] 2025-07-14 21:16:52
Error: Not Found

### Minimal reproduction confirmation

- [x] Yes, I have tested and provided a `minimal.lua` file that reproduces the issue

--- a/lua/codecompanion/adapters/huggingface.lua
+++ b/lua/codecompanion/adapters/huggingface.lua
@@ -2,12 +2,10 @@ local config = require("codecompanion.config")
 local curl = require("plenary.curl")
 local log = require("codecompanion.utils.log")
 local openai = require("codecompanion.adapters.openai")
-local utils = require("codecompanion.utils.adapters")
 
 -- Cache variables for models
 local _cached_models
 local _cache_expires
-local _cache_file = vim.fn.tempname()
 
 ---Get a list of available Hugging Face models from inference providers
 ---@param adapter CodeCompanion.Adapter
@@ -33,12 +31,12 @@ local function get_models(adapter)
     })
   end)
   if not ok then
-    log:error("Could not get Hugging Face models from %s.\nError: %s", url, response)
+    log:error("Could not get Hugging Face models from %s. Error: %s", url, response)
     return {}
   end
-  local ok, models_data = pcall(vim.json.decode, response.body)
-  if not ok then
-    log:error("Error parsing Hugging Face models response: %s", response.body)
+  local ok2, models_data = pcall(vim.json.decode, response.body)
+  if not ok2 then
+    log:error("Error parsing Hugging Face models response from %s", url)
     return {}
   end
   local models = {}
@@ -57,7 +55,7 @@ local function get_models(adapter)
     }
   end
   _cached_models = models
-  _cache_expires = utils.refresh_cache(_cache_file, config.adapters.opts.cache_models_for)
+  _cache_expires = os.time() + (config.adapters.opts.cache_models_for or 1800)
   log:debug("Found %d Hugging Face models", vim.tbl_count(models))
   return models
 end

--- a/lua/codecompanion/adapters/huggingface.lua
+++ b/lua/codecompanion/adapters/huggingface.lua
@@ -1,5 +1,66 @@
+local config = require("codecompanion.config")
+local curl = require("plenary.curl")
 local log = require("codecompanion.utils.log")
 local openai = require("codecompanion.adapters.openai")
+local utils = require("codecompanion.utils.adapters")
+
+-- Cache variables for models
+local _cached_models
+local _cache_expires
+local _cache_file = vim.fn.tempname()
+
+---Get a list of available Hugging Face models from inference providers
+---@param adapter CodeCompanion.Adapter
+---@return table
+local function get_models(adapter)
+  if _cached_models and _cache_expires and _cache_expires > os.time() then
+    return _cached_models
+  end
+  if not adapter.env.api_key then
+    log:error("No Hugging Face API key found")
+    return {}
+  end
+  local url = "https://huggingface.co/api/models?inference_provider=all"
+  local headers = {
+    ["User-Agent"] = "CodeCompanion.nvim",
+  }
+  local ok, response = pcall(function()
+    return curl.get(url, {
+      sync = true,
+      headers = headers,
+      insecure = config.adapters.opts.allow_insecure,
+      proxy = config.adapters.opts.proxy,
+    })
+  end)
+  if not ok then
+    log:error("Could not get Hugging Face models from %s.\nError: %s", url, response)
+    return {}
+  end
+  local ok, models_data = pcall(vim.json.decode, response.body)
+  if not ok then
+    log:error("Error parsing Hugging Face models response: %s", response.body)
+    return {}
+  end
+  local models = {}
+  for _, model_data in ipairs(models_data) do
+    if model_data.id then
+      models[model_data.id] = { opts = { can_stream = true } }
+    end
+  end
+  if vim.tbl_count(models) == 0 then
+    log:warn("No models found from API, using fallback models")
+    models = {
+      ["meta-llama/Llama-3.1-8B-Instruct"] = { opts = { can_stream = true } },
+      ["Qwen/Qwen2.5-32B-Instruct"] = { opts = { can_stream = true } },
+      ["google/gemma-3-27b-it"] = { opts = { can_stream = true } },
+      ["moonshotai/Kimi-K2-Instruct"] = { opts = { can_stream = true } },
+    }
+  end
+  _cached_models = models
+  _cache_expires = utils.refresh_cache(_cache_file, config.adapters.opts.cache_models_for)
+  log:debug("Found %d Hugging Face models", vim.tbl_count(models))
+  return models
+end
 
 ---@class HuggingFace.Adapter: CodeCompanion.Adapter
 return {
@@ -17,18 +78,14 @@ return {
     text = true,
     tokens = false,
   },
-  url = "${url}/models/${model}/v1/chat/completions",
+  url = "https://router.huggingface.co/v1/chat/completions",
   env = {
     api_key = "HUGGINGFACE_API_KEY",
-    url = "https://api-inference.huggingface.co",
-    model = "schema.model.default",
   },
   headers = {
     ["Content-Type"] = "application/json",
     Authorization = "Bearer ${api_key}",
   },
-  -- NOTE: currently, decided to not implement the tokens counter handle, since the API infernce docs
-  -- says it is supported, yet, the usage is returning null when the stream is enabled
   handlers = {
     ---@param self CodeCompanion.Adapter
     ---@return boolean
@@ -39,7 +96,6 @@ return {
       return true
     end,
 
-    --- Use the OpenAI adapter for the bulk of the work
     form_parameters = function(self, params, messages)
       return openai.handlers.form_parameters(self, params, messages)
     end,
@@ -62,14 +118,11 @@ return {
       order = 1,
       mapping = "parameters",
       type = "enum",
-      desc = "ID of the model to use from Hugging Face.",
-      default = "Qwen/Qwen2.5-72B-Instruct",
-      choices = {
-        "meta-llama/Llama-3.2-1B-Instruct",
-        "Qwen/Qwen2.5-72B-Instruct",
-        "google/gemma-2-2b-it",
-        "mistralai/Mistral-Nemo-Instruct-2407",
-      },
+      desc = "ID of the model to use from Hugging Face Inference Providers.",
+      default = "Qwen/Qwen2.5-32B-Instruct",
+      choices = function(adapter)
+        return get_models(adapter)
+      end,
     },
     ---@type CodeCompanion.Schema
     temperature = {
@@ -89,7 +142,7 @@ return {
       mapping = "parameters",
       type = "integer",
       optional = true,
-      default = 2048,
+      default = 4096,
       desc = "The maximum number of tokens to generate.",
       validate = function(n)
         return n > 0, "Must be greater than 0"
@@ -107,26 +160,20 @@ return {
         return n >= 0 and n <= 1, "Must be between 0 and 1"
       end,
     },
-    -- caveat to using the cache: https://huggingface.co/docs/api-inference/parameters#caching
     ---@type CodeCompanion.Schema
-    ["x-use-cache"] = {
+    stop = {
       order = 5,
-      mapping = "headers",
-      type = "string",
+      mapping = "parameters",
+      type = "list",
       optional = true,
-      default = "true",
-      desc = "Whether to use the cache layer on the inference API...",
-      choices = { "true", "false" },
-    },
-    ---@type CodeCompanion.Schema
-    ["x-wait-for-model"] = {
-      order = 6,
-      mapping = "headers",
-      type = "string",
-      optional = true,
-      default = "false",
-      desc = "Whether to wait for the model to be loaded...",
-      choices = { "true", "false" },
+      default = nil,
+      subtype = {
+        type = "string",
+      },
+      desc = "Up to 4 sequences where the API will stop generating further tokens.",
+      validate = function(l)
+        return #l >= 1 and #l <= 4, "Must have between 1 and 4 elements"
+      end,
     },
   },
 }

